### ISSUE TITLE
the case against full html comparison in taglib testing

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#700]
**Reported by @cristianbogdan on 7 Jan 2010 11:17 UTC**
makumba has an interesting mechanism for testing JSP output (i.e. the taglib) which goes something like this (correct me if i'm wrong):
- when the programmer likes the output of a JSP test page, it is recorded in a HTML file. 
- then the test suite compares the output of the test page with the reference HTML
- if there are differences, the test fails and the offending programmer sees a nice DIFF of the output

while this looks like a neat design, it has important shortcomings. i should say from the outset that i never wrote taglib tests from scratch, therefore i may miss important points. nevertheless, from my outside observer perspective:

1) it may be that the HTML differs, but it is still _correct_. for example pointer values for the test records may change during the lifetime of a JVM while the feature tested works perfectly at both database and view level. Another example is a CGI query string where a=A&b=B is equivalent to b=B&a=A but the HTML comparison misses this.

2) most importantly, if the programmer makes a mistake when generating the reference HTML, and "accepts" a HTML which is in fact buggy, the bug will propagate, leading to confusion, because tests pass, don't they? 

3) if a feature different than the one tested is buggy, this will affect all the tests that involve that feature, leading to a larger number of failed tests, thus affecting the programmer morale, besides leading also to confusion about what exactly fails. on the other hand, if there is no specific test for the breaking feature, HTML comparison can help to raise an alarm

this entry is a constructive attempt to find a solution to this problem.

there were two approaches before the present full HTML comparison
1) use markers in the page and compare just between these markers
2) use a library that parses HTML and is capable of extracting body tags (like give me the next tag, i expect it to be a table, from that one give me the first TR, i expect it to contain 2 TDs, and so on)

while i think the first approach (unfortunately rather widespread in the early mak taglib tests) is a bit naive, i think the second approach is quite good, and can be combined with the HTML comparison approach in that it could help us compare just (essential) parts of the page structure, not entire pages. i remember we had a library that could parse HTML quite well and it could even populate and submit forms.

also, i think that even if we manage to replace full html comparison, it will still be useful to have 1-2 large pages that do overall functionality tests, to be able to raise alarms in case of missing individual feature tests

Migrated-From: http://trac.makumba.org/ticket/1155

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 4 Feb 2010 18:07 UTC**
(In reply to comment #0)
> there were two approaches before the present full HTML comparison
> 1) use markers in the page and compare just between these markers
> 2) use a library that parses HTML and is capable of extracting body tags (like
> give me the next tag, i expect it to be a table, from that one give me the
> first TR, i expect it to contain 2 TDs, and so on)
> 
> while i think the first approach (unfortunately rather widespread in the early
> mak taglib tests) is a bit naive, i think the second approach is quite good,
> and can be combined with the HTML comparison approach in that it could help us
> compare just (essential) parts of the page structure, not entire pages. i
> remember we had a library that could parse HTML quite well and it could even
> populate and submit forms.
> 
> also, i think that even if we manage to replace full html comparison, it will
> still be useful to have 1-2 large pages that do overall functionality tests, to
> be able to raise alarms in case of missing individual feature tests

while I do agree with the shortcomings mentioned with the current DIFF (just witnessed its weakness, I think regarding problems-2)), and the method mentioned in solutions-2) seems reasonible, I wonder how we can address this issue from a resource point of view.

I remember that the main problem before switching to HTML diff was that there were not many tests, and I think most of the old tests were written by the belgian students, so eventually I switched to HTML diff in order to increase the amount of automated tests (and that part worked, we now have many more).

further, even a full HTML diff would not be suffiscient to address problems with components in makumba (if we think of the 'ROI' of automated testing).

the problem with testing makumba vs. testing another tag-library is that we can't directly work with the Tag classes, a page context mockup and so forth, because:
- there's page analysis that fundamentally alters the lifecycle of makumba tags vs. normal tags
- there's a filter that works on the request, and even the response

the current automated tests do a full execution of the page, using a real application server, and thus in the end we're only able to do so-called integration tests, rather than components tests or unit tests (there's too much global state in makumba, we have loads of static class members, which simply makes it impossible to do unit testing or even component testing).

----
in conclusion I think that more specific testing would be great, but I also think that currently we are just not enough people coding makumba in order to both still code makumba, and code such precise tests for it.

my reasoning is that the amount of work necessary for coding precise tests  would most probably outweight the amount of work necessary to implement new features (further, maintaining the test suite would also demand a lot of work, should a recurring value - e.g. the form responder value - change during development)

whilst this is fine for systems having many users and a big community (and more coders), I think it makes little sense for makumba with its current lack of functionality and therefore of users.

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:24 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
